src/ck3/data/characters.rs

```diff
@@ -301,125 +301,270 @@ impl Character {
+                    }
+                    "give_council_position" => {
+                        data.verify_exists(Item::CouncilPosition, value);
+                        Some((GiveCouncilPosition, key))
+                    }
+                    _ => {
+                        let msg = format!("unknown key `{key}` in assignment");
+                        err(ErrorKey::Validation).msg(msg).loc(key).push();
+                        None
+                    }
                 }
             }
-        }
-
-        vd.field_item("culture", Item::Culture);
-        if let Some(token) = vd.field_value("set_culture") {
-            validate_prefix_reference_token(token, data, "culture");
-        }
-
-        vd.multi_field_item("trait", Item::Trait);
-        vd.multi_field_item("add_trait", Item::Trait);
-        vd.multi_field_item("remove_trait", Item::Trait);
+            BV::Block(block) => match key.as_str() {
+                "death" => {
+                    let mut vd = Validator::new(block, data);
+                    vd.req_field("death_reason");
+                    vd.field_item("death_reason", Item::DeathReason);
+                    vd.field_item("killer", Item::Character);
+                    Some((Death, key))
+                }
+                "effect" => {
+                    validate_effect(block, data, sc, Tooltipped::No);
+                    None
+                }
+                _ => {
+                    let msg = format!("unknown key `{key}` in definition");
+                    err(ErrorKey::Validation).msg(msg).loc(key).push();
+                    None
+                }
+            },
+        };
 
-        for token in vd.multi_field_value("add_pressed_claim") {
-            validate_prefix_reference_token(token, data, "title");
-        }
-        for token in vd.multi_field_value("remove_claim") {
-            validate_prefix_reference_token(token, data, "title");
-        }
+        result.map(|(etype, token)| (etype, token.clone()))
+    }
 
-        if let Some(token) = vd.field_value("capital") {
-            data.verify_exists(Item::Title, token);
-            if !token.as_str().starts_with("c_") {
-                error(token, ErrorKey::Validation, "capital must be a county");
+    fn validate_life(
+        character: &Token,
+        mut traits: FnvHashSet<Token>,
+        life_events: Vec<LifeEvent>,
+    ) {
+        let mut birth = None;
+        let mut death = None;
+        let mut spouse = None;
+        // does not check for `add_trait` in effect block
+        let mut claims = FnvHashSet::<Token>::default();
+        let mut employed = false;
+
+        for LifeEvent { date, index: _, token, event } in life_events.into_iter() {
+            use LifeEventType::*;
+
+            if birth.is_none() && event != Birth {
+                let msg = format!("{character} is not born yet");
+                let mut loc = token.loc.clone();
+                loc.column = 0;
```

Comment:
Otherwise only the token rather than the line  is highlighted, which can be confusing, even though it is the whole action that is invalid. For example, add_pressed_claim would highlight the title, which could be misleading

Reply:
Tricky :) Like, `remove_claim` should have the claim value highlighted if the problem is that the claim was never added, and it should have the `remove_claim` key highlighted if the problem was that it's happening after death.

I'm still thinking this over. I don't really like the departure from tiger's usual habit of highlighting the tokens. `column = 0` is supposed to be a last resort if no token is available.

Reply:
Yes I have corrected it to use the token for non-birth/death warnings now. They should now point to the relevant claim/trait/etc.